File: test/qr.ml
#use "topfind";;
#require "ctypes";;
#require "ctypes.foreign";;

open Ctypes;;
open Foreign;;

type qrcode
let qrcode : qrcode structure typ = structure "QRcode";;
let _qr_version = field qrcode "version" int;;
let qr_width   = field qrcode "width"   int;;
let _qr_data   = field qrcode "data"    (ptr char);;
let () = seal qrcode;;

let encode_string =
  foreign "QRcode_encodeString"
    (string @-> int @-> int @-> int @-> int @-> returning (ptr qrcode));;

let qr_free =
  foreign "QRcode_free"
    (ptr qrcode @-> returning void);;

let () =
  let qr = encode_string "unsandbox-qr-ok" 0 1 2 1 in
  let w = getf !@qr qr_width in
  Printf.printf "QR:unsandbox-qr-ok:ROWS:%d\n" w;
  qr_free qr;;

File: test/qr.nim
import dynlib

type
  QRcode {.pure.} = object
    version: cint
    width: cint
    data: pointer

type
  QRencodeString = proc(s: cstring, ver: cint, level: cint, hint: cint, cs: cint): ptr QRcode {.cdecl.}
  QRfree = proc(qr: ptr QRcode) {.cdecl.}

let lib = loadLib("libqrencode.so")
if lib == nil:
  quit("Failed to load libqrencode.so", 1)

let encode = cast[QRencodeString](lib.symAddr("QRcode_encodeString"))
let free_qr = cast[QRfree](lib.symAddr("QRcode_free"))

let qr = encode("unsandbox-qr-ok", 0, 1, 2, 1)
if qr != nil:
  echo "QR:unsandbox-qr-ok:ROWS:", qr.width
  free_qr(qr)
else:
  echo "QR encode failed"

unloadLib(lib)
